### PR TITLE
Add transitions to light flash

### PR DESCRIPTION
**Pull request by alexyao2015**
# What does this implement/fix? 
Adds transitions to flash so it adheres to the default transition length set.

Also, fixes flash not restoring state properly. 

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1420

## Test Environment

- [ ] ESP32
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/light/transformers.h [hunk 1]
```diff
@@ -58,7 +58,39 @@ class LightFlashTransformer : public LightTransformer {
...
+      this->transition_length_ = this->length_ / 2;
+  }
+
+  optional<LightColorValues> apply() override {
+    float p = this->get_progress_();
+    if (this->transformer_ != nullptr) {
+      if (!this->transformer_->is_finished()) {
+        return this->transformer_->apply();
+      } else {
+        this->transformer_ = nullptr;
```
Comment (oxan) [inline esphome/components/light/transformers.h:85]: `stop()` should be called here.

File: esphome/components/light/transformers.h [hunk 2]
```diff
@@ -58,7 +58,39 @@ class LightFlashTransformer : public LightTransformer {
...
+      }
+    }
+
+    if (this->last_transition_p_ < p) {
```
Comment (oxan) [inline esphome/components/light/transformers.h]: `last_transition_p_` is used uninitialized.

File: esphome/components/light/transformers.h [hunk 3]
```diff
@@ -58,7 +58,39 @@ class LightFlashTransformer : public LightTransformer {
...
+      if (p < 0.5f) {
+        // first transition to original target
+        this->transformer_ = this->state_.get_output()->create_default_transition();
+        this->transformer_->setup(this->state_.current_values, this->target_values_, this->transition_length_);
+        this->last_transition_p_ = p + 0.5f;
+      } else if (p >= 0.5f && p < 1.0f) {
```
Comment (oxan) [inline esphome/components/light/transformers.h]: This starts the transition back to the start value after 50% of the flash, which will cut the flash short if the transition duration is less than half of the flash length.

**Review by oxan — CHANGES_REQUESTED** [review]
I'm not sure whether unconditionally using the default transition length here is a good idea. It wouldn't surprise me if there are people that want to keep the old behaviour.

File: esphome/components/light/transformers.h (hunk 3 above)
Comment (alexyao2015) [reply esphome/components/light/transformers.h]: Good catch! Thanks.

File: esphome/components/light/transformers.h [hunk 4]
```diff
@@ -58,7 +58,48 @@ class LightFlashTransformer : public LightTransformer {
...
+        this->last_transition_p_ = p + 0.5f;
+        // transformer must be applied immediately here instead of next loop to account for 0 transition length
+        return this->transformer_->apply();
+      } else if (p >= this->transition_back_p_ && p < 1.0f) {
+        // second transition back to start value
+        this->transformer_ = this->state_.get_output()->create_default_transition();
+        this->transformer_->setup(this->state_.current_values, this->get_start_values(), this->transition_length_);
+        this->last_transition_p_ = p + 0.5f;
+        return this->transformer_->apply();
+      }
```
Comment (oxan) [inline esphome/components/light/transformers.h]: This whole logic here seems unnecessarily complicated now. Can we setup the first transition in `start()`, and the second one in a `if (millis() > this->start_time_ + this->length_ - this->transition_length_)`? That way the whole dance with `last_transition_p_` and `transition_back_p_` can be dropped, which is less code and makes it more obvious what's going on as well, imo.

**Review by oxan — COMMENTED** [review]
Now that flashes support transitions, we should support setting the transition length for an individual flash through a `LightCall` as well (though that could be left for a follow-up PR).

File: esphome/components/light/transformers.h (hunk 4 above)
Comment (alexyao2015) [reply esphome/components/light/transformers.h]: There still needs to be a way to mark the secondary transition as having already occurred. I can do that, but I would need another boolean to mark as true once its done or there will be an infinite transition loop.

**Review by oxan — APPROVED** [review]

File: esphome/components/light/transformers.h (hunk 4 above)
Comment (oxan) [reply esphome/components/light/transformers.h]: Why? When the second transition is finished, the flash transformer would be finished, so it'd never start a third transition.

**Pull request merged**